src/input/input_python.rs

```diff
@@ -291,6 +291,15 @@ impl<'py> Input<'py> for Bound<'py, PyAny> {
             return Ok(ValidationMatch::exact(EitherFloat::Py(float.clone())));
         }
 
+        if self.is_instance_of::<PyComplex>() {
+            if strict {
+                return Err(ValError::new(ErrorTypeDefaults::FloatType, self));
+            }
+            let real = self.getattr(intern!(self.py(), "real")).unwrap();
```

Comment:
Perhaps we should raise a warning here. In numpy's case, they do raise a custom `ComplexWarning` for discarding the imaginary part. At least we should document whatever coercion is implemented here because there can be a few possibilities for complex -> float:
* no coercion is allowed: consistent with the default python behaviour, e.g. `float(complex(1, 2))` gives an error, even when the imaginary part is 0.
* coercion is allowed only when the imaginary part is 0: this is what I would actually expect when coercions can actually take place.
* coercion is allowed for all complex numbers, done by simply dropping the imaginary part: this seems to be a numpy-only behaviour and I'm not sure if non-numpy users would expect this.

Reply:
I am somewhat open to lax mode allowing `complex` -> `float` *if the imaginary component is zero*, similarly we allow `datetime` -> `date` *if the datetime is at midnight*.

I think we should never drop a nonzero imaginary part.